addons/godot-next/resources/resource_collections/resource_array.gd

```diff
@@ -15,70 +15,71 @@ class_name ResourceArray
 
 ##### PROPERTIES #####
 
-var _name: = ""
-var _type: Script = null
-
 var _data := []
 
 ##### NOTIFICATIONS #####
 
-func _get_property_list() -> Array:
-	return [ PropertyInfo.new_array("_data", PROPERTY_HINT_NONE, "", PROPERTY_USAGE_STORAGE).to_dict() ]
-
-##### OVERRIDES #####
-
-func can_contain(p_property: String) -> bool:
-	if not _name or not _type:
-		return false
-	return p_property.begins_with(_name + DELIMITER) and not p_property.ends_with(EMPTY_ENTRY)
-
-func add_element(script) -> void:
-	_data.append(script.new())
-
-func set_element(p_property: String, p_value) -> bool:
-	var index := int(p_property.lstrip(_name + DELIMITER + "item_"))
-	if not p_value:
-		_data.remove(index)
-	else:
-		_data[index] = p_value
-	return true
-
-func get_element(p_property: String):
-	var index := int(p_property.lstrip(_name + DELIMITER + "item_"))
-	return _data[index] if index < _data.size() else null
+func _init() -> void:
+	resource_name = COLLECTION_NAME
+
+func _get(p_property: String):
+	if p_property.begins_with(PREFIX):
+		var index := int(p_property.lstrip(PREFIX + "item_"))
+		return _data[index] if index < _data.size() else null
+	return null
+
+func _set(p_property, p_value):
+	if p_property.begins_with(PREFIX):
+		if ClassType.new(p_value).is_type(_type):
+			var index := int(p_property.lstrip(PREFIX + "item_"))
+			if not p_value:
+				_data.remove(index)
+				property_list_changed_notify()
+			else:
+				var res = _instantiate_script(p_value) if p_value is Script else p_value
+				if res:
+					_data[index] = res
+		return true
+	return false
 
-func get_collection_property_list() -> Array:
+func _get_property_list() -> Array:
 	var list = []
-	
-	if not _name or not _type:
+	if not _type:
 		return list
-		
-	var group_prefix = _name + DELIMITER
-	var item_prefix = _name + DELIMITER + "item_"
-	
-	list.append(PropertyInfo.new_group("%s %s" % [_name, COLLECTION_NAME], group_prefix).to_dict())
-	list.append(PropertyInfo.new_array(_name, PROPERTY_HINT_RESOURCE_TYPE, "#%s" % _type.get_path()).to_dict())
 	
+	list.append(PropertyInfo.new_group(PREFIX, PREFIX).to_dict())
+	list.append(PropertyInfo.new_array(PREFIX, PROPERTY_HINT_RESOURCE_TYPE, "#%s" % _type.get_path(), PROPERTY_USAGE_DEFAULT).to_dict())
 	if _data.empty():
-		list.append(PropertyInfo.new_nil("%s%s" % [group_prefix, EMPTY_ENTRY]).to_dict())
-	for idx in _data.size():
-		list.append(PropertyInfo.new_resource("%s%s" % [item_prefix, idx], "", PROPERTY_USAGE_EDITOR).to_dict())
+		list.append(PropertyInfo.new_nil(PREFIX + EMPTY_ENTRY).to_dict())
+	for i in _data.size():
+		list.append(PropertyInfo.new_resource("%sitem_%s" % [PREFIX, i], "", PROPERTY_USAGE_EDITOR).to_dict())
+	
 	return list
 
+##### OVERRIDES #####
+
+func _add_element(script) -> void:
+	_data.append(script.new())
+
+func _refresh_data() -> void:
+	if _type == null:
+		clear()
+		return
+	var data_cache := _data.duplicate()
+	for element in data_cache:
+		if not ClassType.new(element).is_type(_type):
```

Comment:
Rather than constructing new instances of ClassType all the time, we should just keep a private variable in the ResourceArray class itself (and probably ResourceSet since I think it has similar code) and just do `ct.res = element` if `element` is a `Script` or `Resource` object. That will then allow you to re-use the same instance for an arbitrary number of checks, without each new instance re-loading the script map from ProjectSettings.

Reply:
Actually, better to use `ct.path = element.resource_path`.

Reply:
figured it out!